tests/platform_tests/fwutil/conftest.py

```diff
@@ -0,0 +1,107 @@
+import tarfile
+import json
+import pytest
+import logging
+
+from random import randrange
+
+from fwutil_common import show_firmware
+
+logger = logging.getLogger(__name__)
+
+DUT_HOME="/home/admin"
+DEVICES_PATH="/usr/share/sonic/device"
+FS_PATH_TEMPLATE = "/host/image-{}/fs.squashfs"
+FS_RW_TEMPLATE = "/host/image-{}/rw"
+FS_WORK_TEMPLATE = "/host/image-{}/work"
+FS_MOUNTPOINT_TEMPLATE = "/tmp/image-{}-fs"
+OVERLAY_MOUNTPOINT_TEMPLATE = "/tmp/image-{}-overlay"
+
+def check_path_exists(path):
+    return duthost.stat(path = path)["stat"]["exists"] 
+
+def pytest_generate_tests(metafunc):
+    val = metafunc.config.getoption('--fw-pkg')
+    if 'fw_pkg_name' in metafunc.fixturenames and val is not None:
+        metafunc.parametrize('fw_pkg_name', [val], scope="module")
+
+@pytest.fixture(scope='module')
+def fw_pkg(fw_pkg_name):
+    logger.info("Unpacking firmware package to ./firmware")
+    os.mkdir("firmware")
+    with tarfile.open(fw_pkg_name, "r:gz") as f:
+        f.extractall("./firmware/")
+        with open('./firmware/firmware.json', 'r') as fw:
+            fw_data = json.load(fw)
+            yield fw_data
+    subprocess.call("rm -rf firmware", shell=True)
+
+@pytest.fixture(scope='function')
+def random_component(duthost, fw_pkg):
+    chass = show_firmware(duthost)["chassis"].keys()[0]
+    components = fw_pkg["chassis"].get(chass, {}).get("component", []).keys()
+
+    if len(components) == 0:
+        pytest.skip("No suitable components found in config file for platform {}.".format(duthost.facts['platform']))
+
+    return components[randrange(len(components))] 
+
+@pytest.fixture(scope='function')
+def host_firmware(localhost, duthost):
+    logger.info("Starting local python server to test URL firmware update....")
+    comm = "python3 -m http.server --directory {}".format(os.path.join(DEVICES_PATH, 
+        duthost.facts['platform']))
+    task, res = duthost.command(comm, module_ignore_errors=True, module_async=True)
+    yield "http://localhost:8000/"
+    task.terminate()
+
+@pytest.fixture(scope='function')
+def next_image(duthost, fw_pkg):
+
+    # Install next version of sonic
+    current = duthost.shell('sonic_installer list | grep Current | cut -f2 -d " "')['stdout']
+
+    image = fw_pkg.get("images", {}).keys()
```

Comment:
@alexrallen this is not a part of your test plan (https://github.com/Azure/sonic-mgmt/blob/master/docs/testplan/FWUtil-test-plan.md) and looks a bit of overhead, since sonic image is not a platform firmware. Why not to use a command line option?

Reply:
This allows for flexibility, the vendor may include whatever image they want to here. This is a part of my test plan here https://github.com/Azure/sonic-mgmt/blob/master/docs/testplan/FWUtil-test-plan.md#432-successful-update-from-next-image

I don't want to force the end user to have to host the image somewhere if they want to use a specific image for this. 

Compared to firmware installations this is actually very little overhead with regard to time (other than file size but I don't see a good reason this is limited)

Reply:
@alexrallen thanks for pointing out, but still, it's not a part of package description.
How do you expect user to write the valid config file in that case?

Also, why do we need a two different versions?
```json
{
        "images":{
                "master.153-d99595a0e": "images/SONiC-OS-master.153-d99595a0e.bin",
                "master.154-edb48e319": "images/SONiC-OS-master.154-edb48e319.bin"
        },
}
```

Reply:
We need two different images for the same reason we need two different firmware versions, in case it aligns with the installed image. A vender could have as few or as many images as they want here though.

Reply:
I have not updated the HLD because I have to submit a pull request to the community every time I update it. I am currently collecting all of the changes and will update the HLD in a single pull request once all review of this code is done and I am confident the schema for the JSON file will not change.